### PR TITLE
Tweaks to ItemsSourceView

### LINKED ISSUE #5571: ComboBox indices are not updated well when bound collection is updated
**Describe the bug**
My combobox is bound to a collection that implemets INotifyCollectionChanged. if I clear the collection, add items more than the previous count, the combobox menu updates with all the new items, and the old ones cleared as expected. But selecting any of the items which indices exceed the last index in the previous collection will set the selected index to -1 and clear the box text, even though the item exist in the collection and appears in the menu

**To Reproduce**
1. Create a Combobox, in code or Xaml
2. Set the Items property to an observerable collection with items
3. Clear the items, and add more items than what was in before
4. Select any of the new items with indices greater that the last item in the previous collection
5. Selected Index is set to -1 

**Expected behavior**
Selected Index should be set to the correct index

**Desktop (please complete the following information):**

- OS: Windows 10 20H2
- Version: 0.10.0

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

A few tweaks/fixes for `ItemsSourceView`:

- Lazy subscribe to inner collection changed: subscribing to the inner collection's `CollectionChanged` event when an `ItemsSourceView` was created meant the only way to unsubscribe is to dispose the `ItemsSourceView` meaning that the instance can't be easily shared
- Validate source collection
  - Don't allow wrapping an existing `ItemsSourceView` - should use the `GetOrCreate` static method if this is a possibility
  - Don't allow source collections which implement `INotifyCollectionChanged` but not `IList` - this leads to hard-to-diagnose problems like #5571 
- Throw when accessing a disposed `ItemsSourceView`

## Fixed issues

Closes #5571

### CODE REVIEW

File: src/Avalonia.Controls/ItemsSourceView.cs [hunk 1]
```diff
@@ -42,27 +42,22 @@ public class ItemsSourceView : INotifyCollectionChanged, IDisposable
...
-                _inner = new List<object>(objectEnumerable);
-            }
-            else
+            _inner = source switch
             {
-                _inner = new List<object>(source.Cast<object>());
-            }
-
-            ListenToCollectionChanges();
+                ItemsSourceView => throw new ArgumentException("Cannot wrap an existing ItemsSourceView.", nameof(source)),
```
Comment (Takoooooo) [inline src/Avalonia.Controls/ItemsSourceView.cs:47]: This feature is not supported on our SDK, please update

Comment (Takoooooo) [reply src/Avalonia.Controls/ItemsSourceView.cs:47]: image

Comment (grokys) [reply src/Avalonia.Controls/ItemsSourceView.cs:47]: Balls. We really need to update our SDK.

Comment (grokys) [reply src/Avalonia.Controls/ItemsSourceView.cs]: Fixed.

File: src/Avalonia.Controls/ItemsSourceView.cs [hunk 2]
```diff
@@ -82,25 +90,60 @@ public ItemsSourceView(IEnumerable source)
...
+        public event NotifyCollectionChangedEventHandler? CollectionChanged
+        {
+            add
+            {
+                if (_inner is null)
+                    ThrowDisposed();
+
+                if (_collectionChanged is null)
+                {
+                    if (_inner is INotifyCollectionChanged incc)
```
Comment (Takoooooo) [inline src/Avalonia.Controls/ItemsSourceView.cs:102]: Can we write 
``` csharp
if (_collectionChanged is null&&_inner is INotifyCollectionChanged incc) 
```
 ?

File: src/Avalonia.Controls/ItemsSourceView.cs [hunk 3]
```diff
@@ -82,25 +90,60 @@ public ItemsSourceView(IEnumerable source)
         /// <summary>
         /// Occurs when the collection has changed to indicate the reason for the change and which items changed.
         /// </summary>
-        public event NotifyCollectionChangedEventHandler? CollectionChanged;
+        public event NotifyCollectionChangedEventHandler? CollectionChanged
+        {
+            add
+            {
+                if (_inner is null)
+                    ThrowDisposed();
```
Comment (Takoooooo) [inline src/Avalonia.Controls/ItemsSourceView.cs:98]: Can we reference ```Inner``` below instead of that check?

**Review by Takoooooo — APPROVED** [review]

**Comment by Takoooooo** [comment]
LGTM. Tested code in the issue with cibuild from this PR, works well. But I am curious why we don't get here the adequate name of source
image

File: src/Avalonia.Controls/ItemsSourceView.cs (hunk 3 above)
Comment (grokys) [reply src/Avalonia.Controls/ItemsSourceView.cs]: Good points! Fixed.

**Comment by grokys** [comment]
Yeah, that's strange @Takoooooo! Looks to be a problem with VS? https://github.com/dotnet/runtime/issues/29502

**Comment by Takoooooo** [comment]
Yes, i think so.
image
image
But the fact that they are not fixing it for years is sad.

**Comment by Kermalis** [comment]
I'm sorry but why do you need to implement IList, rather than IReadOnlyList?

Now I'm forced to stay on an old version because IList is completely inappropriate for many use cases and I can't change collections from other libraries lol. Just adds more bloat to anything Avalonia will ever be used with. I get you need it to solve your bug but there has to be another way (IReadOnlyList is just an indexer guys). And until then I'm staying on 10.0.5 because of it.

**Comment by maxkatz6** [comment]
@Kermalis FYI, IList doesn't inherit IReadOnlyList. So if develop will pass collection that implements IList, but not IReadOnlyList, it won't work.
But I can agree, that it should be possible to support both. @grokys

**Pull request merged**